Mathlib/MeasureTheory/Measure/SeparableMeasure.lean

```diff
@@ -0,0 +1,486 @@
+        generateSetAlgebra_mono <| subset_union_left ..
+    · induction hs with
+      | base t t_mem =>
+        rcases t_mem with t_mem | ⟨n, rfl⟩
+        · exact hgen ▸ measurableSet_generateFrom t_mem
+        · exact μ.toFiniteSpanningSetsIn.set_mem n
+      | empty => exact MeasurableSet.empty
+      | compl t _ t_mem => exact MeasurableSet.compl t_mem
+      | union t u _ _ t_mem u_mem => exact MeasurableSet.union t_mem u_mem
+
+/-- If a measurable space is countably generated and equipped with an s-finite measure, then the
+measure is separable. -/
+instance [CountablyGenerated X] [SFinite μ] : IsSeparable μ where
+  exists_countable_measureDense := by
+    have := isSeparable_of_sigmaFinite (μ := μ.restrict μ.sigmaFiniteSet)
+    rcases exists_countable_measureDense (μ := μ.restrict μ.sigmaFiniteSet) with ⟨𝒜, count_𝒜, h𝒜⟩
+    let ℬ := {s ∩ μ.sigmaFiniteSet | s ∈ 𝒜}
+    refine ⟨ℬ, count_𝒜.image (fun s ↦ s ∩ μ.sigmaFiniteSet), ?_, ?_⟩
+    · rintro - ⟨s, s_mem, rfl⟩
+      exact (h𝒜.measurable s s_mem).inter (measurableSet_sigmaFiniteSet μ)
+    · intro s ms hμs ε ε_pos
+      rcases restrict_compl_sigmaFiniteSet_eq_zero_or_top μ s with hs | hs
+      · have : (μ.restrict μ.sigmaFiniteSet) s ≠ ∞ :=
+          ne_top_of_le_ne_top hμs <| μ.restrict_le_self _
+        rcases h𝒜.approx s ms this ε ε_pos with ⟨t, t_mem, ht⟩
+        refine ⟨t ∩ μ.sigmaFiniteSet, ⟨t, t_mem, rfl⟩, ?_⟩
+        rw [← measure_inter_add_diff _ (measurableSet_sigmaFiniteSet μ)]
+        have : μ (s ∆ (t ∩ μ.sigmaFiniteSet) \ μ.sigmaFiniteSet) = 0 := by
+          rw [diff_eq_compl_inter, inter_symmDiff_distrib_left, ← ENNReal.bot_eq_zero, eq_bot_iff]
+          calc
+            μ ((μ.sigmaFiniteSetᶜ ∩ s) ∆ (μ.sigmaFiniteSetᶜ ∩ (t ∩ μ.sigmaFiniteSet)))
+              ≤ μ ((μ.sigmaFiniteSetᶜ ∩ s) ∪ (μ.sigmaFiniteSetᶜ ∩ (t ∩ μ.sigmaFiniteSet))) :=
+                measure_mono symmDiff_subset_union
+            _ ≤ μ (μ.sigmaFiniteSetᶜ ∩ s) + μ (μ.sigmaFiniteSetᶜ ∩ (t ∩ μ.sigmaFiniteSet)) :=
+                measure_union_le _ _
+            _ = 0 := by
+                rw [inter_comm, ← μ.restrict_apply ms, hs, ← inter_assoc, inter_comm, ← inter_assoc,
+                  inter_compl_self, empty_inter, measure_empty, zero_add]
+        rwa [this, add_zero, inter_symmDiff_distrib_right, inter_assoc, inter_self,
+          ← inter_symmDiff_distrib_right, ← μ.restrict_apply' (measurableSet_sigmaFiniteSet μ)]
+      · refine False.elim <| hμs ?_
+        rw [eq_top_iff, ← hs]
+        exact μ.restrict_le_self _
+
+end IsSeparable
+
+section SecondCountableLp
+
+/-! ### A sufficient condition for $L^p$ spaces to be second-countable -/
+
+/-- If the measure `μ` is separable (in particular if `X` is countably generated and `μ` is
+`s`-finite), if `E` is a second-countable `NormedAddCommGroup`, and if `1 ≤ p < +∞`,
+then the associated `Lᵖ` space is second-countable. -/
+instance [IsSeparable μ] [SecondCountableTopology E] : SecondCountableTopology (Lp E p μ) := by
```

Comment:
Is this better? In some sense, it's morally a weaker assumption (even though they are equivalent given the other assumptions on `E`), especially because type class inference only has an instance in the second countable → separable direction.
```suggestion
instance [IsSeparable μ] [TopologicalSpace.SeparableSpace E] : SecondCountableTopology (Lp E p μ) := by
```

Reply:
You're absolutely right!